packages/cli/src/models/network/ProjectDeployer.ts

```diff
@@ -0,0 +1,137 @@
+import _ from 'lodash';
+import { AppProject, PackageProject, SimpleProject, App, Package, ImplementationDirectory } from 'zos-lib';
+
+import NetworkController from './NetworkController';
+import { ZosPackageFile, ZosNetworkFile } from '../files/index';
+
+interface PartialDeploy {
+  app?: App;
+  thepackage?: Package;
+  directory?: ImplementationDirectory;
+}
+
+interface ExistingAddresses {
+  appAddress?: string;
+  packageAddress?: string;
+}
+
+interface CreateProjectFn {
+  (addresses: ExistingAddresses): Promise<AppProject>;
+}
+
+class BaseProjectDeployer {
+  protected controller: NetworkController;
+  protected packageFile: ZosPackageFile;
+  protected networkFile: ZosNetworkFile;
+  protected txParams: any;
+  protected requestedVersion: string;
+
+  constructor(controller: NetworkController, requestedVersion: string) {
+    this.controller = controller;
+    this.packageFile = controller.packageFile;
+    this.networkFile = controller.networkFile;
+    this.txParams = controller.txParams;
+    this.requestedVersion = requestedVersion;
+  }
+}
+
+class BasePackageProjectDeployer extends BaseProjectDeployer {
+  get packageAddress(): string {
+    return this.controller.packageAddress;
+  }
+
+  protected _tryRegisterPartialDeploy({ thepackage, directory }: PartialDeploy): void {
+    if (thepackage) this._registerPackage(thepackage);
+    if (directory) this._registerVersion(this.requestedVersion, directory);
+  }
+
+  protected _registerPackage({ address }: { address: string }): void {
```

Comment:
Probably not worth doing anything atm, but note that `{ address }` is the de-structuring of an incoming object that has a property "address" (and could have any other properties). OTOH, the typing `{ address: string }` is a type that defines an object that _only_ has the property "address". 

I'm not a 100% sure about this, but let's just keep this in the back of our heads for now.

Reply:
Already discussed it offline.